cloudinit/net/dhcp.py

```diff
@@ -503,62 +551,275 @@ def get_latest_lease(lease_dir, lease_file_regex):
+            dhcp request
+        @param dhcp_log_func: A callable accepting the client output and
+            error streams.
+        @param distro: a distro object for network interface manipulation
+        @return: dict of lease options representing the most recent dhcp lease
+            parsed from the dhclient.lease file
+        """
+        LOG.debug("Performing a dhcp discovery on %s", interface)
+        sleep_time = 0.01
+        sleep_cycles = int(self.timeout / sleep_time)
+
+        # dhcpcd needs the interface up to send initial discovery packets
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        distro.net_ops.link_up(interface)
+        try:
+            command = [
+                self.dhcp_client_path,  # pyright: ignore
+                "--ipv4only",  # only attempt configuring ipv4
+                "--waitip",  # wait for ipv4 to be configured
+                "--persistent",  # don't deconfigure when dhcpcd exits
+                "--noarp",  # don't be slow
+                "--script=/bin/true",  # disable hooks
+                interface,
+            ]
+            out, err = subp.subp(
+                command,
+                timeout=self.timeout,
+            )
+            if dhcp_log_func is not None:
+                dhcp_log_func(out, err)
+            lease = self.get_newest_lease(interface)
+            # Attempt cleanup and leave breadcrumbs if it fails, but return
+            # the lease regardless of failure to clean up dhcpcd.
+            if lease:
+                # pid file arrival can take very long, and stdout might contain
+                # the PID, so attempt to use a stdout parsed PID
+                for line in reversed(out.split("\n")):
+                    if "forked to background, child pid " in line:
+                        try:
+                            pid = int(line.split()[-1])
+                            gid = distro.get_proc_pgid(pid)
+                            if gid:
+                                os.kill(gid, signal.SIGKILL)
+                                break
+                        except ValueError:
+                            LOG.debug(
+                                "Clouldn't kill process: couldn't parse "
+                                "pid from stdout"
+                            )
+                        except ProcessLookupError:
+                            LOG.debug(
+                                "Couldn't kill process, it already exited"
+                            )
+                LOG.debug("PID not in stdout, waiting for PID file")
+
+                # Note: the pid file location depends on the arguments passed
+                # it can be discovered with the -P flag
+                pid_file = subp.subp([*command, "-P"]).stdout
```

Comment:
Also we need to `strip()` this response otherwise we are always hitting issues with `load_file(/run/dhcpcd/*pid\n")` being unable to find the file. This is what resulted in the 1000 retries trying to determine pids on ec2 for me.

Reply:
> subp shell out is a bit heavier an operation that just glob("/run/dhcpcd/<interface_name>.pid")

Also I'm not convinced that such a glob would actually be faster - when this command runs, dhcpcd is already in page cache and just requires a fork/exec (of the already-cached binary) then a few instructions. A glob would require `stat`ing the directory, and if the directory isn't already in the dcache then it probably means a round trip to disk and back. I agree that subp is heavy-weight - but I wouldn't assume that it's faster than globbing directories in the case that the binary is already in memory and only a couple of syscalls away from a returned result.

> /run/dhcpcd/<interface_name>.pid

I didn't document it, but I'm pretty sure I saw dhcpcd put the lease in a different directory or filename before. That's why I did this at all, given the manpage documentation of the lease location. I was curious about the existence of `-P` at all - until I saw that happen.

Reply:
> Also we need to strip() this response otherwise we are always hitting issues with load_file(/run/dhcpcd/*pid\n") being unable to find the file. This is what resulted in the 1000 retries trying to determine pids on ec2 for me.

Thanks for finding that. I did test that this worked - the original `strip()` must have gotten lost when I rebased for the review.